ast/policy.go

```diff
@@ -148,11 +157,12 @@ type (
 	// Rule represents a rule as defined in the language. Rules define the
 	// content of documents that represent policy decisions.
 	Rule struct {
-		Location *Location `json:"-"`
-		Default  bool      `json:"default,omitempty"`
-		Head     *Head     `json:"head"`
-		Body     Body      `json:"body"`
-		Else     *Rule     `json:"else,omitempty"`
+		Location    *Location    `json:"-"`
+		Default     bool         `json:"default,omitempty"`
+		Head        *Head        `json:"head"`
+		Body        Body         `json:"body"`
+		Else        *Rule        `json:"else,omitempty"`
+		Annotations []Annotation `json:"annotation,omitempty"`
```

Comment:
I'm not sure we want to store the annotations directly on rules. For example, suppose we wanted to annotate the entire file or package to indicate the structure of the input document. If we store the annotations on the module we can still look them up from the rule (because it has a pointer to the module that contains it.)

Reply:
@tsandall That is a very good reason to move Annotations! 

One other case I'm wondering though is the scope higher than package/module we want to support where we annotate with schemas at the `global` level (via a `schema-config.yaml` file in the `schemas` directory or something to that effect). At this scope, do we add the global schema annotations to ALL the Modules{}, that will then be referenced by every single rule? Or is there a better place for global annotations to take effect?

Reply:
@aavarghese that's a good question.

I think we can separate how users supply annotations that have global scope versus how we implement things internally. If we put the annotations on the module, we are creating coupling between modules and annotations (which would make it harder to support an external `schema-config.yaml` file) but I'm not sure that's a problem. The alternative would be to store the annotations outside of modules and then provide them separately to components that require them (e.g., the compiler, evaluator, other tooling, etc.)

What are the most obvious annotation/schema scopes?

- rule-level (what this PR is addressing immediately)
- file-level
- package-level (multiple .rego files can contribute to the same file)
- hierarchical, e.g., "apply this schema to all rules in package a.b.c as well as child packages (a.b.c.d, a.b.c.d.e.f, etc.)"
- global

Storing the annotations on the module will solve the rule-level and file-level scopes. For the others, we have to do analysis across modules and additional data structures have to be created. So far, this is typically done in the compiler. I think we could address the package-level/hierarchical/global use cases there as well...sorry for rambling...

I think we can support the package-level, hierarchical, and global use cases by including a field on the annotation that specifies intent (we could use a package naming convention?) I'd be inclined to keep the schema annotations in the .rego files rather than a separate YAML file as it will be easier to be integrated tooling.

Thoughts?

Reply:
Two immediate (or few hours late actually) thoughts or questions on this @tsandall...

1. What if we stored rule level annotations in Rule{} (with `type: rule`) and file-level annotations in Module{} (with `type: module`)? This way, if a rule finds itself having a schema annotation, it has it in `Rule.Annotation` and if it doesn't have a schema annotation, it can check its Module{} `Rule.Module.Annotation` to see if there is a file-level annotation that will need to be applied. Otherwise, each and every rule will have to do a search through `Module.Annotation[]` array to find its annotation (which might be a search on Rule.Head.Name stored in annotation) even the rules that don't have an annotation!! Affect performance?

Sorry for discussing implementation in a post like this. My feeling is that it could be neater/more organized letting Rule store its own Annotation[], and Module its own Annotation[]. Maybe I'm missing the real advantage of having the annotations for ALL Rules stored in one place in its Module...

Other than that, I'm in agreement with you on the different higher scopes, and how we can analyze them across modules in the  compiler etc. 💯  👍 

2. Regarding global annotations in .rego files vs a separate yaml file, tooling integration is a sure pro. My concern would be mainly for our repos that have a few hundreds of Rego files and how we can globally annotate them. Adding annotation in every file isn't a practical expectation from our users.

Reply:
My thought process around putting annotations on the ast.Module struct was that we're going to need additional data structures to support the hierarchical and global use cases anyway. Rather than storing the annotations on both ast.Module and ast.Rule, I'd rather just store them all on ast.Module and then build the data structures to provide lookups for the different use cases we have (e.g., lookup for rule, lookup for file, lookup w/ hierarchical scope, etc.) I think it's a bit of a tradeoff...it's easier right now to store annotations on the ast.Rule

Reply:
@aavarghese thinking about this a bit more...let's just go ahead with the annotation at the rule level. It should unblock the PR significantly and we can iterate during this release if needed.

Reply:
@tsandall Points raised above are very valid. After the weekend, decided to move Annotations into Module struct, like you suggested. Better to set the foundation for this metadata feature in this PR, instead of redoing it all in the next.

So we have Annotations now in Module being set in [getAnnotation](https://github.com/open-policy-agent/opa/pull/3123/files#diff-b9267ea1f38ea7d750c9feb543ae391e6b1be9c1fd06125c769ae86443e59ed8R106), and then the array being traversed to find the annotation corresponding to the rule being checked in [checkRule](https://github.com/open-policy-agent/opa/pull/3123/files#diff-d23ff3bfce1aa610d34183f8ccab5ecfd112e19dd9ebca55d69f231a29b62337R258) and schemas used to set typeenv appropriately.

Commit for this work is merged into the one big commit (1bb648291462bc31bbb397b55ce19b7404efeae3). Please let us know your feedback/